common/src/main/java/com/wynntils/core/webapi/WebManager.java

```diff
@@ -286,34 +295,35 @@ public static CompletableFuture<Boolean> tryLoadMaps() {
                         String md5 = mapObject.get("hash").getAsString();
 
                         // TODO DownloaderManager? + Overlay
-                        handler.addRequest(new RequestBuilder(fileBase + file, file)
+                        WynntilsMod.info("Starting downloading map piece: " + file);
+                        handler.addAndDispatch(new RequestBuilder(fileBase + file, file)
                                 .cacheTo(new File(mapDirectory, file))
                                 .cacheMD5Validator(md5)
                                 .useCacheAsBackup()
                                 .handle(bytes -> {
                                     try (ByteArrayInputStream in = new ByteArrayInputStream(bytes)) {
                                         NativeImage nativeImage = NativeImage.read(in);
 
-                                        syncList.add(new MapProfile(file, nativeImage, x1, z1, x2, z2));
+                                        downloadedMapProfiles.add(new MapProfile(file, nativeImage, x1, z1, x2, z2));
                                     } catch (IOException e) {
                                         WynntilsMod.info("IOException occurred while loading map image of " + file);
                                         return false; // don't cache
                                     }
 
+                                    WynntilsMod.info("Successfully downloaded map piece: " + file);
                                     return true;
                                 })
                                 .build());
                     }
 
-                    // hacky way to know when handler has finished dispatching
-                    CompletableFuture.runAsync(handler::dispatch).whenComplete((a, b) -> {
-                        if (syncList.size() == mapArray.size()) {
-                            result.complete(true);
-                            maps = syncList;
-                        } else {
-                            result.complete(false);
-                        }
-                    });
+                    if (downloadedMapProfiles.size() == mapArray.size()) {
+                        maps = downloadedMapProfiles;
+                        result.complete(true);
+                    } else {
+                        WynntilsMod.error("Expected to have" + mapArray.size() + " map parts, got "
+                                + downloadedMapProfiles.size() + " parts.");
+                        result.complete(false);
+                    }
```

Comment:
The reason for the hack is that having added a bunch of map download requests done asynchronously, we need to do an operation when all are complete.

Now, we are doing the whole map download synchronously, which likely dramatically increases startup time

Reply:
Yeah, but can't that make a race somewhere? As we do not group requests, and only dispatch 1 by 1

Reply:
I feel the solution here is to add better grouping functionality/use the existing grouping functionality and code that runs when the entire group is done by default.